presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestDriver.java

```diff
@@ -1264,6 +1292,71 @@ public void testBadUrlMissingSchema()
         }
     }
 
+    @Test(timeOut = 10000)
+    public void testQueryCancellation()
+            throws Exception
+    {
+        try (Connection connection = createConnection("blackhole", "blackhole");
+                Statement statement = connection.createStatement()) {
+            statement.executeUpdate("CREATE TABLE test_cancellation (key BIGINT) " +
+                    "WITH (" +
+                    "   split_count = 1, " +
+                    "   pages_per_split = 1, " +
+                    "   rows_per_page = 1, " +
+                    "   page_processing_delay = '10s'" +
+                    ")");
+        }
+
+        CountDownLatch queryStarted = new CountDownLatch(1);
+        CountDownLatch queryFinished = new CountDownLatch(1);
+        AtomicReference<String> queryId = new AtomicReference<>();
+        AtomicReference<Throwable> queryFailure = new AtomicReference<>();
+
+        Future<?> queryFuture = executorService.submit(() -> {
+            try (Connection connection = createConnection("blackhole", "default");
+                    Statement statement = connection.createStatement();
+                    ResultSet resultSet = statement.executeQuery("SELECT * FROM test_cancellation")) {
+                queryId.set(resultSet.unwrap(PrestoResultSet.class).getQueryId());
+                queryStarted.countDown();
+                try {
+                    resultSet.next();
+                }
+                catch (SQLException t) {
+                    queryFailure.set(t);
+                }
+                finally {
+                    queryFinished.countDown();
+                }
+            }
+            return null;
+        });
+
+        queryStarted.await(10, SECONDS);
+        assertNotNull(queryId.get());
+        assertFalse(getQueryState(queryId.get()).isDone());
+        queryFuture.cancel(true);
+        queryFinished.await(10, SECONDS);
+        assertNotNull(queryFailure.get());
+        assertEquals(getQueryState(queryId.get()), FAILED);
+
+        try (Connection connection = createConnection("blackhole", "blackhole");
+                Statement statement = connection.createStatement()) {
+            statement.executeUpdate("DROP TABLE test_cancellation");
+        }
+    }
+
+    private QueryState getQueryState(String queryId)
```

Comment:
Instead of using `QueryState` which is part of the engine, I would just return the string. It also avoids the enum `valueOf` failing if, for whatever reason, a non-valid value is returned (the `assertEquals` would fail then and give a good error message).